sfnt/table_feature.go

```diff
@@ -0,0 +1,417 @@
+	// langSysRecords[langSysCount] langSysRecord // Array of LangSysRecords, listed alphabetically by LangSys tag
+}
+
+type featureTable struct {
+	FeatureParams    uint16 // = NULL (reserved for offset to FeatureParams)
+	LookupIndexCount uint16 // Number of LookupList indices for this feature
+	// lookupListIndices [lookupIndexCount]uint16 // Array of indices into the LookupList — zero-based (first lookup is LookupListIndex = 0)}
+}
+
+type lookupTable struct {
+	Type          uint16 // Different enumerations for GSUB and GPOS
+	Flag          uint16 // Lookup qualifiers
+	SubTableCount uint16 // Number of subtables for this lookup
+	// subtableOffsets[subTableCount] uint16 // Array of offsets to lookup subtables, from beginning of Lookup table
+	// markFilteringSet uint16 // Index (base 0) into GDEF mark glyph sets structure. This field is only present if bit useMarkFilteringSet of lookup flags is set.
+}
+
+type langSysTable struct {
+	LookupOrder          uint16 // = NULL (reserved for an offset to a reordering table)
+	RequiredFeatureIndex uint16 // Index of a feature required for this language system; if no required features = 0xFFFF
+	FeatureIndexCount    uint16 // Number of feature index values for this language system — excludes the required feature
+	// featureIndices[featureIndexCount] uint16 // Array of indices into the FeatureList, in arbitrary order
+}
+
+// parseLangSys parses a single Language System table. b expected to be the beginning of Script table.
+// See https://www.microsoft.com/typography/otspec/chapter2.htm#langSysTbl
+func (t *TableLayout) parseLangSys(b []byte, record langSysRecord) (*LangSys, error) {
+	if int(record.Offset) >= len(b) {
+		return nil, io.ErrUnexpectedEOF
+	}
+
+	r := bytes.NewReader(b[record.Offset:])
+
+	var lang langSysTable
+	if err := binary.Read(r, binary.BigEndian, &lang); err != nil {
+		return nil, fmt.Errorf("reading langSysTable: %s", err)
+	}
+
+	featureIndices := make([]uint16, lang.FeatureIndexCount, lang.FeatureIndexCount)
+	if err := binary.Read(r, binary.BigEndian, &featureIndices); err != nil {
+		return nil, fmt.Errorf("reading langSysTable featureIndices[%d]: %s", lang.FeatureIndexCount, err)
+	}
+
+	var features []*Feature
+	for i := 0; i < len(featureIndices); i++ {
+		if int(featureIndices[i]) >= len(t.Features) {
+			return nil, fmt.Errorf("invalid featureIndices[%d] = %d", i, featureIndices[i])
+		}
+		features = append(features, t.Features[featureIndices[i]])
+	}
+
+	return &LangSys{
+		Tag:      record.Tag,
+		Features: features,
+	}, nil
+}
+
+// parseScript parses a single Script table. b expected to be the beginning of ScriptList.
+// See https://www.microsoft.com/typography/otspec/chapter2.htm#sTbl_lsRec
+func (t *TableLayout) parseScript(b []byte, record scriptRecord) (*Script, error) {
+	if int(record.Offset) >= len(b) {
+		return nil, io.ErrUnexpectedEOF
+	}
+
+	b = b[record.Offset:]
```

Comment:
@bramp Is there a byte length on the record? I'm worried about the parsing code reading beyond the length of the record.

Reply:
nope, there is sadly no length for any of these fields. I suspect they are all contained within the single table.

Reply:
ah yes, all the records are within the same table, and that b array only contains the current table:

```go
buffer := io.NewSectionReader(file, int64(entry.Offset), int64(entry.Length))
table, err := parseTable(entry.Tag, buffer)
...
    b, err := ioutil.ReadAll(buffer)
...
    b = b[record.Offset:]